### ISSUE TITLE
Phyloseq and metacoder

### ISSUE THREAD

**Issue by wipperman** [grunwaldlab/metacoder#141]
Can you point me to a tutorial or accessor function for how to take an OTU table + metadata and convert this to a taxmap object, which can then be used to make plots? Is this possible with the package? I see that it is in the future directions in the paper, but am unable to figure it out myself (although am able to run all of the software and the examples!). Thanks so much for the help.

**Comment by zachary-foster** [comment]
Hi @wipperman! So you have a phyloseq object then? There is no function yet to handle this yet, but I can probably show you how to do it. I plan on adding this, but I am working on rewriting the `taxmap` class in the [taxa](https://github.com/ropenscilabs/taxa) repo, so a lot of upgrades are waiting on that. Once `taxmap` class is more flexible, it will make a lot of these compatibility issues easy to fix. 

In the mean time, can you send me an example of how your data is formatted (an .Rdata file) or the code to produce it or a link to a phyloseq tutorial that produces an object with the same format? I can probably figure out how to do the conversion and perhaps add the function to the dev branch of metacoder. Thanks

**Comment by wipperman** [comment]
Sure! Here is some reproducible code that uses sample data from the Phyloseq package. The code will plot a tree on the left, with a differential abundance heatmap on the right, sorted by bodysite. I would like to use your package to show what genera (at any level of taxonomy, Order, Genus, etc) differ between just two of the groups (e.g., skin and soil). As for writing a wrapper function, it is this kind of question that I believe many users of Phyloseq would like to so (compare treatment vs control), so this may be a helpful example to incorporate since it uses data available through Phyloseq. Let me know if you have any other questions @zachary-foster and thanks so much for the help! 

rm (list = ls())
library(phyloseq);library(ggstance);library(ggthemes);library(plyr);library(dplyr);library(ggtree);library(tibble)

set.seed(394582)

#load sample Phyloseq dataset
data("GlobalPatterns")
GP <- GlobalPatterns
GP <- prune_taxa(taxa_sums(GP) > 500, GP)
humantypes <- c("Feces", "Skin") #make it a bit simpler 
sample_data(GP)$human <- get_variable(GP, "SampleType") %in% humantypes

mergedGP <- merge_samples(GP, "SampleType")
mergedGP <- rarefy_even_depth(mergedGP,rngseed=394582)
mergedGP <- tax_glom(mergedGP,"Order") #make the tree smaller for the purposes of this example

print(mergedGP) #final phyloseq object

#make a tree
tr <- phy_tree(mergedGP)
spec <- tax_table(mergedGP) %>% data.frame(stringsAsFactors = FALSE) %>% tibble::rownames_to_column("otu")
gt <- ggtree(tr, branch.length = "y") %<+% spec 
gd <- gt$data

melt_data<-psmelt(mergedGP) 

tt <- melt_data %>%left_join(select(gd,OTU=label,x,y),by="OTU") %>%
  arrange(Sample) %>% mutate(sample2=factor(Sample,levels=unique(Sample)),
  col=as.numeric(sample2),x.col=scales::rescale(col,to=c(1.3,2)))

#this will demonstrate that one can make a ggtree object with abundance data (other examples show this tpp)
g <- gt + geom_tippoint(data=gd$istip,aes(color=Phylum),size=2) + 
  geom_text(data=gd$istip,aes(label=Genus,x=x+0.01,size='smaller[1]'),hjust=0.1,check_overlap = T)+ geom_tile(data=tt,aes(x=x.col,y=y,fill=Sample,alpha=Abundance)) + theme(legend.position="right") #+ geom_tiplab2()
g

**Comment by zachary-foster** [comment]
Hi  @wipperman, sorry about the delay. Last week was busy. 

I have managed to make a less-than-elegant function to convert phyloseq objects to `taxmap` objects for use with metacoder. It will probably be in the dev version shortly. In the mean time, I have pasted it below and an example of how to use it. The code below assumes you have run the code you supplied above and that you are using the dev version of metacoder. This function will probably be rewritten in the next major release once the [taxa](https://github.com/ropenscilabs/taxa) package (where `taxmap` will live in the future) is on CRAN.

```r
# install dev version of metacoder
devtools::install_github("grunwaldlab/metacoder@dev")
library(metacoder)

# First draft of conversion function 
phyloseq_to_taxmap <- function(phyloseq_obj) {
  tax_data <- data.frame(
    taxonomy = vapply(seq_len(nrow(phyloseq_obj@tax_table)), FUN.VALUE = character(1),
                      function(i) {
                        taxon_names <- as.vector(phyloseq_obj@tax_table[i, ])
                        rank_names <- colnames(phyloseq_obj@tax_table)
                        rank_names <- rank_names[!is.na(taxon_names)]
                        taxon_names <- taxon_names[!is.na(taxon_names)]
                        paste(rank_names, taxon_names, sep = "::", collapse = ";;")
                      }),
    phyloseq_id = rownames(phyloseq_obj@tax_table)
  )
  
  otu_data  <- as.data.frame(t(phyloseq_obj@otu_table))
  otu_data$phyloseq_id <- rownames(otu_data)
  tax_data <- merge(tax_data, otu_data, by.x = "phyloseq_id", by.y = "phyloseq_id")
  
  parse_taxonomy_table(tax_data, taxon_col = c("class" = 2), other_col_type = "obs_info",
                       class_regex = "^(.*)::(.*)$",
                       class_key = c(rank = "taxon_info", "name"),
                       class_sep = ";;")
}

# Convert to taxmap
data <- phyloseq_to_taxmap(mergedGP)

# Get proportions for each taxon from OTU counts
sample_names <- rownames(mergedGP@sam_data)
data$taxon_data[sample_names]  <- lapply(sample_names, function(col_name) {
  vapply(obs(data), function(indexes) sum(data$obs_data[indexes, col_name]) / sum(data$obs_data[, col_name]), numeric(1))
})

# Test plot of difference between ocean and soil
set.seed(2)
heat_tree(data,
          node_size = rowMeans(cbind(Ocean, Soil)),
          node_label = name,
          tree_label = name,
          node_color = Ocean - Soil,
          node_color_range = diverging_palette(),
          node_color_interval = c(-0.5, 0.5),
          edge_color_interval = c(-0.5, 0.5),
          node_label_max = 200,
          layout = "da", initial_layout = "re",
          title = "Ocean vs Soil",
          node_size_axis_label = "Average proportion")
```

![rplot02](https://cloud.githubusercontent.com/assets/6090555/25158424/2e0ca0ce-245d-11e7-9fb9-3205bc615f92.png)


Feel free to ask any questions if something isn't clear.

**Comment by wipperman** [comment]
@zachary-foster thanks so much! I am able to reproduce it with my example. One oversight on my part is that OTU tables are usually structured by samples. In the example I gave, the work of merging samples by some variable (ocean, soil, etc) was done for you in the OTU table. A function that takes the OTU table and metadata, and merges many samples that fall into some category for a comparison (healthy/treatment, ocean/soil, etc) would be more helpful. It would add about one line of code, taking the metadata of the phyloseq object and merging samples based off of this metadata. 

I can send another example as a .R file with some different data if you want to give it a try, or need other data to work with. I am not able to get this function to run on my own--I'm having issues merging the otu_table and tax_table using the merge function (it is deleting my data). I'll keep trying and let you know how it goes. But thanks for doing this--these trees will add a lot to the easy visualization of microbiome data comparisons.

**Comment by zachary-foster** [comment]
No problem @wipperman! 

> . A function that takes the OTU table and metadata, and merges many samples that fall into some category for a comparison (healthy/treatment, ocean/soil, etc) would be more helpful.

Yea, that is the data structure I am more used to seeing. One of the reasons I am making taxmap more flexible is to make encoding multiple samples per treatment easier to standardize so I can make functions that can interpret this information automatically. What I do currently is just have all samples for all treatments in one of the tables and a separate (i.e. not in the taxmap object) "metadata" table that has which samples are in which treatments. In the future, this metadata table will be in the taxmap object, like it is in phyloseq's object. 

> . I am not able to get this function to run on my own...

Im not sure what you mean, but I can help if you send me a part of your data. 

Thanks

**Comment by wipperman** [comment]
Hey @zachary-foster 

Here is a sample file where the sample data slot has "status" with two populations of people (healthy vs treatment). This is about as generic as one could get, and represents how people generally interact with phyloseq-class data. Github will only allow me to upload publically if its extension is .zip--just unarchive the file and substitute the phyloseq object like this: 

phy <- readRSD("~/PATH/TO/example.phyloseq")

[example.phyloseq.zip](https://github.com/grunwaldlab/metacoder/files/949045/example.phyloseq.zip)

Hope this helps, and let me know if you have any questions! 
--Matt

**Comment by wipperman** [comment]
Hey @zachary-foster 
Just wondering if you were able to view this/get it to load into R? Is there anything else I can help you with? Let me know if you have any other questions! 
--Matt

**Comment by zachary-foster** [comment]
Hi @wipperman, 

Yea, thanks, I got it to load in R, but my draft parser does not work with it for some reason. I need to look into it more. 

Is your work/research waiting on getting your data into metacoder from phyloseq? If so, I will prioritize getting it to work for you quickly. However, I will soon be reworking the parsing tools of metacoder to handle this kind of thing much better so any quick fixes now will likely be overwritten in the near future anyway.   

The new version of `taxmap` that will be implemented in the [taxa](https://github.com/ropensci/taxa) package should be able to hold all of the information on the phyloseq object easily. We plan to submit taxa to CRAN this month. After that, I will rewrite the file parsers / converters to use this new object and it will greatly simplify a lot of these issues, which is why I have not worked much in the past few weeks on the current parsers/converters.

Sorry if you are waiting on this for your work. If you need/want your data parsed soon, I can help you with your specific case and get it done quick likely. If you are more concerned with having an all-purpose phyloseq converter, that might be best put off until the taxa package is on CRAN in a few weeks.

**Comment by wipperman** [comment]
Hey @zachary-foster!
Thank you for the help--it is not a big rush, I wanted to use the graphics in a talk that I am going to give at the end of the month, but if you give it a shot and are unable to get it to work, I'm able to wait and hold off until there is a phyloseq_to_taxmap-type function in your taxa package that I can use. I'll be sure to keep an eye out for it and adapt it. If you do manage to come up with a quick fix to your parser function in the meantime that would accommodate the datatype from the phyloseq object that I sent, please let me know. Thanks! 
Best, --Matt

**Comment by zachary-foster** [comment]
This is how I might parse phyloseq objects once the [taxa](https://github.com/ropensci/taxa) package is integrated:

```
> parse_tax_data(as.data.frame(mergedGP@tax_table), as.data.frame(t(mergedGP@otu_table)), class_cols = 1:4, mappings = c("{{name}}" = "{{name}}"))
<Taxmap>
  212 taxa: 1. Archaea, 2. Bacteria, 3. Crenarchaeota ... 209. Gemellales, 210. Turicibacterales, 211. Lactobacillales, 212. Bacillales
  212 edges: NA->1, NA->2, 1->3, 1->4, 2->5, 2->6, 2->7, 2->8 ... 79->204, 79->205, 79->206, 79->207, 80->208, 81->209, 81->210, 81->211, 81->212
  1 data sets:
    [[1]]:
      # A tibble: 131 x 10
        taxon_id Feces Freshwater `Freshwater (creek)`  Mock Ocean `Sediment (estuary)`  Skin  Soil Tongue
           <chr> <dbl>      <dbl>                <dbl> <dbl> <dbl>                <dbl> <dbl> <dbl>  <dbl>
      1       82     4          1                    2    19   343                21166     9   278      8
      2       83     1          0                    1    14     1                  378   134  9069      1
      3       84     1         14                 2765     1    15                   10     0     0      2
      # ... with 128 more rows
  0 functions:
```

This is still being refined, but it seems to work in the dev version of [taxa](https://github.com/ropensci/taxa). @wipperman: I am mostly putting this here for my notes, but I thought you might be interested as well.

**Comment by wipperman** [comment]
Hey @zachary-foster, I am definitely still interested--let me download the dev version of taxa, and I'll give it a shot/see what I can do. If you have some sample code from the example phyloseq object I gave you, let me know. Additionally, if I get something to work, I'll post it on here. Thanks so much!

**Comment by zachary-foster** [comment]
Cool, thanks @wipperman! However, the functions in metacoder are not yet adapted to the new `taxmap` class defined in `taxa`. I am working on adapting them now. I would be happy to have your input on the `taxa` package, but the plotting done by `metacoder` will not work with this class yet fyi.

**Comment by wipperman** [comment]
Gotcha--no problem. I've already downloaded the dev version of taxa so I might as well try to learn what you're doing with it. When it's all ready let me know! Thank you @zachary-foster!

**Comment by zachary-foster** [comment]
Hi @wipperman, I have made a draft of the new phylseq object parser that uses the taxa package. I also updated the `heat_tree` function to use this object. There is still more work needed to get things stable enough for a release, but the fundamentals are there.

Previously you said: 

>  A function that takes the OTU table and metadata, and merges many samples that fall into some category for a comparison (healthy/treatment, ocean/soil, etc) would be more helpful. 

I agree, but I have not done that yet. Should not be hard though, assuming all phyloseq objects have sample data and otu tables in the same format.

You can try out what I am working on with the following code:

```r
devtools::install_github("grunwaldlab/metacoder")
devtools::install_github("ropensci/taxa")

library(metacoder)
library(taxa)

phy <- readRDS("example.phyloseq")
obj <- parse_phyloseq(phy)

obj$filter_taxa(taxon_names == "Bacteria", subtaxa = TRUE)

healthy_samples <- obj$data$sam_data$sample_id[obj$data$sam_data$status == "healthy"]
healthy_counts <- obs_apply(obj, "otu_table", function(i) sum(obj$data$otu_table[i, healthy_samples]), simplify = TRUE)

obj %>%
  heat_tree(node_size = healthy_counts, 
            node_color = healthy_counts,
            node_label = taxon_names)
```

Let me know if you see any changes you would like to see to the object returned by `parse_phyloseq`.

**Comment by wipperman** [comment]
Hey @zachary-foster ...this looks great to me. I've attached the output from your code, and think it looks great! 

One idea may be to allow for two continuous color scales to be plotted onto the same tree. So when comparing healthy vs treatment, the differences are more obvious (it's almost like a Circos plot but with an actually phylogenetic tree). Another suggestion is to put in the help section of the "parse_phyloseq" function this example, or something like it, so that people can understand how to go from parsing to plotting. 

I'll play around with it some more, but at a first pass, this is exactly what I had in mind! Thanks so much!

![healthy_taxmap](https://user-images.githubusercontent.com/16925318/28752278-aa6ea78e-751b-11e7-98e1-d1ec37053ec6.jpeg)

**Comment by zachary-foster** [comment]
Thanks for trying it out @wipperman! Im glad you like it so far

> One idea may be to allow for two continuous color scales to be plotted onto the same tree.

That can kind of be done by setting the `edge_color` to something different than the `node_color`. I am also planning on making "piecharts" of each of the nodes when more than one value is given per taxon, so that might be an option in the future. You can also compare two samples by subtracting/dividing the values for each taxon, as was done in the ocean vs soil plot above. 

> Another suggestion is to put in the help section of the "parse_phyloseq" function this example, or something like it, so that people can understand how to go from parsing to plotting.

Good idea, I will do that.

**Comment by wipperman** [comment]
@zachary-foster So if you wanted to plot both healthy and treatment on the same tree using your above example, how would this be done? Something like this is what I have in mind. I fiddled with changing the edge_color, but although this gave two scales, I don't think it was right. Thanks! 

healthy_samples <- obj$data$sam_data$sample_id[obj$data$sam_data$status == "healthy"]
healthy_counts <- obs_apply(obj, "otu_table", function(i) sum(obj$data$otu_table[i, healthy_samples]), simplify = TRUE)

treat_samples <- obj$data$sam_data$sample_id[obj$data$sam_data$status == "treatment"]
treat_counts <- obs_apply(obj, "otu_table", function(i) sum(obj$data$otu_table[i, treat_samples]), simplify = TRUE)

obj %>%
  heat_tree(node_size = c(treat_counts,healthy_counts), 
            node_color = c(treat_counts,healthy_counts),
            node_label = taxon_names)

**Comment by zachary-foster** [comment]
Hi @wipperman, sorry for the delay! I was traveling when I noticed this issue and forgot to respond until now. 

Here is how I typically compare two treatments. This code also uses three new functions: `calc_obs_props`, `calc_taxon_abund`, and `compare_treatments` to make this kind of thing easier. I think `compare_treatments` should address you previous suggestion:

>  A function that takes the OTU table and metadata, and merges many samples that fall into some category for a comparison (healthy/treatment, ocean/soil, etc) would be more helpful. 

You will have to install the most recent dev version again.

```r
devtools::install_github("grunwaldlab/metacoder")
devtools::install_github("ropensci/taxa")

library(metacoder)

phy <- readRDS("example.phyloseq")
obj <- parse_phyloseq(phy)

# Convert counts to proportions
obj$data$otu_table <- calc_obs_props(obj,
                                     dataset = "otu_table",
                                     cols = obj$data$sam_data$sample_ids)

# Calculate per-taxon proportions 
obj$data$tax_table <- calc_taxon_abund(obj, 
                                       dataset = "otu_table", 
                                       cols = obj$data$sam_data$sample_ids)

# Calculate difference between treatments
obj$data$diff_table <- compare_treatments(obj,
                                          dataset = "tax_table",
                                          sample_ids = obj$data$sam_data$sample_ids,
                                          treatments = obj$data$sam_data$status)

# Plot differneces
color_interval <- c(-5, 5) # The range of values (log 2 ratio of median proportion) to display
obj %>%
  taxa::filter_taxa(taxon_names == "Bacteria", subtaxa = TRUE) %>%
  heat_tree(node_size_axis_label = "Number of OTUs",
            node_size = n_obs,
            node_color_axis_label = "Log 2 ratio of median proportions",
            node_color = log2_median_ratio,
            node_color_range = diverging_palette(),
            node_color_trans = "linear",
            node_color_interval = color_interval,
            edge_color_interval = color_interval,
            node_label = taxon_names,
            node_label_max = 150)
```

![rplot](https://user-images.githubusercontent.com/6090555/29393646-da7c01be-82b9-11e7-9f53-1ea9eac842d6.png)

Green == more abundant in healthy
Brown == more abundant in treatment

**Comment by wipperman** [comment]
Hey @zachary-foster --this has been working great for me so far! One additional question that you may want to add to the tutorial page as well is how to determine which color corresponds to which "Treatment". So in the above example, Green is more abundant in healthy and Brown more abundant in Treatment, but for other comparisons, how is the color scheme determined? Is it alphabetical? How can I determine this? Thanks!

**Comment by zachary-foster** [comment]
Hi @wipperman, glad to hear it!

Good idea, I will look into this. It depends on the function used. With the default function, it is "sample_1" - "sample_2", with the order of samples the same as they appear in "obj$data$diff_table" is the above example . If the user defines their own function, I cannot determine what the numbers mean however

**Comment by wipperman** [comment]
@zachary-foster Thanks! I think this will confuse people (or at least it will confuse me!) so adding it to either the output of the function or the help section would be useful. 

So just to be clear: using your function "compare_treatments", the output tibble has headers: "taxon_id, treatment_1, treatment_2, log2_median_ratio, median_diff, mean_diff, wilcox_p_value". 

The color closest to Node 1 corresponds to treatment_2, and the color closer to the last node (n=# of OTUs) is treatment_1? For me, "treatment_1" is "healthy" and "treatment_2" is "treatment".

**Comment by zachary-foster** [comment]
> I think this will confuse people (or at least it will confuse me!) so adding it to either the output of the function or the help section would be useful.

I agree, I will add something about this to the documentation.

> The color closest to Node 1 corresponds to treatment_2, and the color closer to the last node (n=# of OTUs) is treatment_1? For me, "treatment_1" is "healthy" and "treatment_2" is "treatment".

Thats right. Its `treatmeant_1` - `treatment_2`, so a positive number indicates more in `treatment_1`.

**Comment by jarrodscott** [comment]
Hello @zachary-foster 

I am attempting to follow the "tutorial" in this thread. any reason I should get `Error: could not find function "compare_treatments"`? Could this possibly have anything to do with my data? Seem unlikely...

I have tried installing the devtools and stable versions. current version of metacoder installed is `‘0.2.0’`
Maybe this has been deprecated in favor of `compare_groups`? However `compare_treatments` looks like a brand new function. Anyway thanks!

**Comment by zachary-foster** [comment]
Hello @jjscarpa. Yes, `compare_groups` is the name I decided on when I released the package to CRAN a few days ago. It used to be called `compare_treatments`. Its the same function with different option names. I changed it to make it more consistent with the other functions.

Take a look at the readme on this repo for a more up to date tutorial: 

https://github.com/grunwaldlab/metacoder

**Comment by jarrodscott** [comment]
awesome thanks for the info. Really looking into checking out this
analytical approach. Super cool. Thanks again

**Comment by zachary-foster** [comment]
No problem, thanks!

**Comment by Marieag** [comment]
Hi! 

I am also very much interested in using metacoder for my project, I've got a phyloseq object I'd like to investigate, but I also hit a snag trying to run your function, I get a `could not find function "parse_taxonomy_table"` .
I would be very happy if you could help me work around it? 

I'm running `metacoder v0.2.0.9007`, just downloaded, with R 3.4.3 and newly installed packages, `vegan v2.4.6` and `phyloseq v1.22.3`. 

My dataset was constructed in Qiime, not mothur, so I'm not sure what to do, if I can't transform it directly from the phyloseq object. Any help would be very much appreciated! 

Thanks!

**Comment by zachary-foster** [comment]
Hello @Marieag, are you using the `phyloseq_to_taxmap` that I posted above a while ago? If so, that has been replaced with `parse_phyloseq` included in the metacoder package (try `?metacoder::parse_phyloseq`). I did find a bug in that parser shortly after the last CRAN release, so I suggest you install the dev version:

```
devtools::install_github("ropensci/taxa")
devtools::install_github("grunwaldlab/metacoder")
```

You are getting that error because `metacoder::parse_taxonomy_table` has been replaced by `taxa::parse_tax_data`.

**Comment by Marieag** [comment]
Hi, thanks for your reply! 

I was indeed using your `phyloseq_to_taxmap` . I reinstalled metacoder and managed to get a taxmap-object using `parse_phyloseq`, but when trying to use it to make heat trees, I get the error 

```
Error in if (zero_range(as.numeric(limits))) { : 
  missing value where TRUE/FALSE needed
In addition: Warning messages:
1: In FUN(X[[i]], ...) :
  Could not apply layout 'reingold-tilford' to subgraph. Using 'fruchterman-reingold' instead.
2: In min(element_data$y) : no non-missing arguments to min; returning Inf

```

Any idea what might cause this?

**Comment by zachary-foster** [comment]
Hmm, hard to say without seeing your object. Can you use `save` to send me the phyloseq object and the code you use to parse and plot it? Thanks

**Comment by Marieag** [comment]
Of course! I should have added that in the last comment. 

[Overall.zip](https://github.com/grunwaldlab/metacoder/files/1697221/Overall.zip)

I'm using the following code (let me know if you need more than this, I'm assuming you'll only need the `metacoder` code?)

```
###-----------Metacoder----------

#parse phyloseq object Overall, created from Qiime .biom-file and table of metadata. 
obj <- parse_phyloseq(Overall)

#removing counts of less than 1
obj$data$otu_table <- zero_low_counts(obj, "otu_table", min_count = 1)

no_reads <- rowSums(obj$data$otu_table[,2:39]) == 0
sum(no_reads)

obj <- filter_obs(obj, "tax_data", ! no_reads, drop_taxa = TRUE)
print(obj)

# Convert counts to proportions
obj$data$otu_table <- calc_obs_props(obj,
                                     dataset = "otu_table",
                                     cols = obj$data$sam_data$sample_ids)

# Calculate per-taxon proportions 
obj$data$tax_table <- calc_taxon_abund(obj, 
                                       dataset = "otu_table", 
                                       cols = obj$data$sam_data$sample_ids)
#construct heat tree
heat_tree(obj, 
          node_label = taxon_names,
          node_size = n_obs,
          node_color = Type)

# Calculate difference between treatments
obj$data$diff_table <- compare_groups(obj, daaset = "tax_table",
                                      cols = obj$data$tax_table$taxon_id,
                                      groups = obj$data$sam_data$Type)

print(obj$data$diff_table)

heat_tree(obj, 
          node_label = taxon_names,
          node_size = n_obs,
          node_color = Type, 
          node_size_axis_label = "OTU count",
          node_color_axis_label = "Samples with reads")

```
I'm getting the same error with both heat tree attempts, and also when running this bit of code on the  GlobalPatterns dataset.
I'm probably missing something really obvious - thank you so much for helping me, I really appreciate it!

**Comment by xpingli** [comment]
`obj <- filter_obs(obj, "tax_data", ! no_reads, drop_taxa = TRUE)` change the "tax_data" to "otu_table";
'# Calculate per-taxon proportions 
obj$data$tax_table <- calc_taxon_abund(obj, 
                                       dataset = "otu_table", 
                                       cols = obj$data$sam_data$sample_ids)' change $tax_table to $tax_abund


This is what I used to mine as I was following this issue. It worked for me but I am not sure if this would help your case.

**Comment by zachary-foster** [comment]
Thanks for the input @xpingli!

@Marieag, sorry for not getting back to you; I forgot about this and @xpingli's reply reminded me. I just looked at it and the problem is that you are trying to use "Type" for node color. There are two reasons this does not work: 

* "Type" is categorical and metacoder currently does not technically support categorical variables, although we are working on it.
* "Type" is from the "sam_data" table, which does not have taxon IDs, so no information from this table can be used directly in plottting. Only information that is associated with taxa can be plotted.

I will try to make the error message more informative though. 

Are you trying to compare "Control" and  "Exclosure"? How about something like: 

```r
load("Overall.RData")

library(metacoder)

obj <- parse_phyloseq(Overall)

# Convert counts to proportions
obj$data$otu_table <- calc_obs_props(obj,
                                     dataset = "otu_table",
                                     cols = obj$data$sam_data$sample_ids)

# Calculate per-taxon proportions 
obj$data$tax_table <- calc_taxon_abund(obj, 
                                       dataset = "otu_table", 
                                       cols = obj$data$sam_data$sample_ids)

# construct heat tree
heat_tree(obj, 
          node_label = taxon_names,
          tree_label = taxon_names,
          node_size = n_obs,
          node_color = n_obs,
          node_size_axis_label = "OTU count",
          layout = "da", initial_layout = "re")

# Calculate difference between treatments
obj$data$diff_table <- compare_groups(obj, dataset = "tax_table",
                                      cols = obj$data$sam_data$sample_ids,
                                      groups = obj$data$sam_data$Type)

# comapre treatments
heat_tree(obj, 
          node_label = taxon_names,
          tree_label = taxon_names,
          node_size = n_obs,
          node_color = log2_median_ratio, 
          node_color_interval = c(-3, 3),
          node_color_range = c("cyan", "gray", "magenta"),
          node_size_axis_label = "OTU count",
          node_color_axis_label = "Log 2 ratio of median proportions",
          layout = "da", initial_layout = "re")
```

![rplot](https://user-images.githubusercontent.com/6090555/38107014-9d0a684c-3345-11e8-82a6-b858e8265573.png)

You might want to reinstall the dev versions of metacoder and taxa.

If you want to plot just the control or encolosure read counts, try using the `cols` and `groups` options in `calc_taxon_abund` (`?calc_taxon_abund`)

**Comment by tarunaaggarwal** [comment]
Hi @zachary-foster,

I'm trying to utilize the commands within this thread but I can't even get past `parse_phyloseq`. I have a merged Phyloseq object, and when I try to parse this object using the parse function, I'm getting `Error: Column taxon_id must be a 1d atomic vector or a list`. I found [this](https://stackoverflow.com/questions/47094086/column-project-must-be-a-1d-atomic-vector-or-a-list?rq=1&utm_medium=organic&utm_source=google_rich_qa&utm_campaign=google_rich_qa) online but I'm not quite sure how to implement this solution to a phyloseq object.

My code so far is below (it's not much). I'm using Phyloseq 1.22.3 and just installed Metacoder yesterday so the version is 0.2.0.9012. Thank you for your help and time in advance!

```
library(phyloseq)
library(taxa)
library(metacoder)
print(hmp_otus)
print(hmp_samples)

setwd("~/Dropbox/bik_lab/github/Funginomicon/memb-files")

# Import mapping file into phyloseq
map <- import_qiime_sample_data("18S-euk-QIIME-mapping-MEmicrobiome-FINAL-26Jul17.txt")

# Import biom tables into phyloseq
otu <- import_biom("otu_table_mc2_w_tax_BlankOTUsRemoved_BlankSamplesRemoved.biom")

#merge mapping file and otu tables into a phyloseq object 
physeq = merge_phyloseq(map,otu)
tax_table(physeq)

#check the otu and tax tables
otu_table(physeq)[1:5]
tax_table(physeq)[1:5]


#parse the phyloseq object for metacoder  
physeq_mc = parse_phyloseq(physeq)


heat_tree(physeq_mc,
          node_size = n_obs,
          node_color = n_obs,
          node_label = taxon_names,
          tree_label = taxon_names)

```
> Output from `sesionInfo()`:
```
R version 3.4.2 (2017-09-28)
Platform: x86_64-apple-darwin15.6.0 (64-bit)
Running under: macOS High Sierra 10.13.4

Matrix products: default
BLAS: /System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libBLAS.dylib
LAPACK: /Library/Frameworks/R.framework/Versions/3.4/Resources/lib/libRlapack.dylib

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] metacoder_0.2.0.9012 taxa_0.2.1           phyloseq_1.22.3     

loaded via a namespace (and not attached):
 [1] reshape2_1.4.3      splines_3.4.2       lattice_0.20-35     rhdf5_2.22.0       
 [5] colorspace_1.3-2    stats4_3.4.2        mgcv_1.8-22         utf8_1.1.3         
 [9] survival_2.41-3     rlang_0.2.0         pillar_1.2.1        glue_1.2.0         
[13] BiocGenerics_0.24.0 bindrcpp_0.2        foreach_1.4.4       bindr_0.1          
[17] plyr_1.8.4          stringr_1.3.0       zlibbioc_1.24.0     Biostrings_2.46.0  
[21] munsell_0.4.3       gtable_0.2.0        codetools_0.2-15    Biobase_2.38.0     
[25] permute_0.9-4       IRanges_2.12.0      biomformat_1.6.0    parallel_3.4.2     
[29] Rcpp_0.12.16        scales_0.5.0        vegan_2.5-1         S4Vectors_0.16.0   
[33] jsonlite_1.5        XVector_0.18.0      ggplot2_2.2.1       stringi_1.1.7      
[37] dplyr_0.7.4         grid_3.4.2          ade4_1.7-8          cli_1.0.0          
[41] tools_3.4.2         magrittr_1.5        lazyeval_0.2.1      tibble_1.4.2       
[45] cluster_2.0.6       crayon_1.3.4        ape_5.1             pkgconfig_2.0.1    
[49] MASS_7.3-47         Matrix_1.2-12       data.table_1.10.4-3 assertthat_0.2.0   
[53] rstudioapi_0.7      iterators_1.0.9     R6_2.2.2            multtest_2.34.0    
[57] igraph_1.2.1        nlme_3.1-131        compiler_3.4.2 
```
Corresponding mapping file and biom table are [here](https://www.dropbox.com/sh/narv17v5z7vxahf/AABEURI2FmWAEuZPvaesAETQa?dl=0).

**Comment by zachary-foster** [comment]
Hello @tarunaaggarwal, sorry for the trouble.

I have not seen that error before. Can you send me the result of `sessionInfo()` and the object using `save(my_obj, file = "my_file_name.RData")`? Its hard for me to diagnose without reproducing the error. Thanks!

**Comment by tarunaaggarwal** [comment]
Hi @zachary-foster  - Sorry I posted the question before I was ready. I'm working on getting you the files. I will edit my original post. Just need a couple more mins

**Comment by tarunaaggarwal** [comment]
Hi @zachary-foster. Did my files go through okay?

**Comment by zachary-foster** [comment]
I don't see them here, did you email them to me?

**Comment by tarunaaggarwal** [comment]
They are at the bottom of my original post as a link.

**Comment by zachary-foster** [comment]
Oh, I see them now. Sorry about that. I was expecting to get an email alert, but I guess they don't send out email alerts for recently edited posts. Thanks for the files! I will look at this today.

**Comment by zachary-foster** [comment]
@tarunaaggarwal, 

I think I have it fixed. It was caused `parse_phyloseq` not knowing what to do with the arbitrary rank names in the phyloseq object. I have not seen that before, so I removed the requirements on rank names and it seems to work. You will need to reinstall:

```r
devtools::install_github("ropensci/taxa")
devtools::install_github("grunwaldlab/metacoder")
```

This works for me:

```r
library(phyloseq)
library(taxa)
library(metacoder)

# Import mapping file into phyloseq
map <- import_qiime_sample_data("18S-euk-QIIME-mapping-MEmicrobiome-FINAL-26Jul17.txt")

# Import biom tables into phyloseq
otu <- import_biom("otu_table_mc2_w_tax_BlankOTUsRemoved_BlankSamplesRemoved.biom")

# merge mapping file and otu tables into a phyloseq object 
physeq = merge_phyloseq(map,otu)

# parse the phyloseq object for metacoder  
physeq_mc = parse_phyloseq(physeq)

# ALTERNATIVE: Parse with cleaner taxon names
#physeq_mc = parse_phyloseq(physeq,
#                           class_regex = "^D?_?[0-9]*_?_?(.+)$")


# Plot
heat_tree(physeq_mc,
          node_size = n_obs,
          node_color = n_obs,
          node_label = taxon_names,
          tree_label = taxon_names)
```

Note that there is a parser for the QIIME biome format called `parse_qiime_biom`. This also works: 

```r
library(metacoder)

# Using the biome parser in metacoder
physeq_mc = parse_qiime_biom("otu_table_mc2_w_tax_BlankOTUsRemoved_BlankSamplesRemoved.biom")

# ALTERNATIVE: Using the biome parser in metacoder with cleaner taxon names
#physeq_mc = parse_qiime_biom("otu_table_mc2_w_tax_BlankOTUsRemoved_BlankSamplesRemoved.biom",
#                             class_regex = "^D?_?[0-9]*_?_?(.+)$")

# Plotting
heat_tree(physeq_mc,
          node_size = n_obs,
          node_color = n_obs,
          node_label = taxon_names,
          tree_label = taxon_names)
```

**Comment by tarunaaggarwal** [comment]
Hi @zachary-foster - thank you very much for working on this but unfortunately, I'm still getting errors (different ones now though). I tried both of your parsing commands after reinstalling the repos and  I think these errors are related to installation :-/ Please see the commands and errors below. 

```
devtools::install_github("ropensci/taxa")
devtools::install_github("grunwaldlab/metacoder")
library(phyloseq)
library(taxa)
library(metacoder)

# Import mapping file into phyloseq
map <- import_qiime_sample_data("18S-euk-QIIME-mapping-MEmicrobiome-FINAL-26Jul17.txt")

# Import biom tables into phyloseq
otu <- import_biom("otu_table_mc2_w_tax_BlankOTUsRemoved_BlankSamplesRemoved.biom")

#merge mapping file and otu tables into a phyloseq object 
physeq = merge_phyloseq(map,otu)

#parse the phyloseq object for metacoder  
physeq_mc = parse_phyloseq(physeq)
```
gives the following error

```
Error in getExportedValue(pkg, name) : 
  lazy-load database '/Library/Frameworks/R.framework/Versions/3.4/Resources/library/taxa/data/Rdata.rdb' is corrupt
In addition: Warning message:
In getExportedValue(pkg, name) : internal error -3 in R_decompress1
```

AND

```
physeq_mc = parse_qiime_biom("otu_table_mc2_w_tax_BlankOTUsRemoved_BlankSamplesRemoved.biom")

```

gives a similar error.

```
Error in get(obj) : 
  lazy-load database '/Library/Frameworks/R.framework/Versions/3.4/Resources/library/taxa/data/Rdata.rdb' is corrupt
In addition: Warning messages:
1: In strsplit(msg, "\n") : input string 1 is invalid in this locale
2: In get(obj) : internal error -3 in R_decompress1
```

**Comment by zachary-foster** [comment]
Yea, I see those occasionally. Try restarting R/Rstudio. That fixes it when it happens to me. I think it is caused by reinstalling packages that are currently loaded.

**Comment by tarunaaggarwal** [comment]
Lovely! Thank you so much! Now, I can finally try what others have posted in this issue thread :-) Thank you, @zachary-foster!

**Comment by zachary-foster** [comment]
No problem! Im glad it worked

**Comment by tarunaaggarwal** [comment]
Hi @zachary-foster - thank you for your help earlier this week. I ran into another issue that I was hoping to get your help with please. I have a biom table that I parse using your alternative approach and this table contains only ONE sample. Perhaps that is why the `rowSums` is giving an error but I'm not sure.  
 > Code
```
obj = parse_qiime_biom("otu_table_mc2_w_tax_BlankOTUsRemoved_BlankSamplesRemoved_nem46.biom", 
                       class_regex = "^D?_?[0-9]*_?_?(.+)$")


obj$data$otu_table <- zero_low_counts(obj, "otu_table", min_count = 5)
no_reads <- rowSums(obj$data$otu_table[, obj$sample_id]) == 0
```

> `rowSums` is giving the error below.

```
Error: Unsupported index type: NULL
```

I put this biom table file in the same folder that I shared with you via Dropbox. THANKS!

**Comment by zachary-foster** [comment]
Hi @tarunaaggarwal, I think it is because `obj$sample_id` does not exist. In the example, there was a table called `hmp_samples` that contained the sample IDs, which were the column names of the abundance matrix. Since you only have one sample,

```r
no_reads <- obj$data$otu_table[, 1] == 0
```

should work the same. rowSums is not needed. There are probably other parts of the code you will need to change as well, since you dont have a sample data table and just have one sample.

**Comment by tarunaaggarwal** [comment]
Ah I see. Lovely! Thanks, @zachary-foster. It is so nice how responsive you are.

**Comment by tarunaaggarwal** [comment]
Hey @zachary-foster - I forgot to ask you a question. Can Metacoder make graphs for each sample individually? I have a biom table for OTUs associated with single nematodes as samples and if I wish to make a graph for each worm separately, is there an easy way to do that in Metacoder? Also, can I make such separate graphs for a particular taxon such as Fungi only?

**Comment by zachary-foster** [comment]
> Can Metacoder make graphs for each sample individually? 

It does not do it automatically, but its quite easy to do. You can put the heat_tree command in a loop / apply function and change the column name for the counts you want to graph in each iteration. Something like:

```r
plot_one <- function(my_col_name) {
  heat_tree(obj,
    node_color = obj$data$my_table[[my_col_name]],
    output_file = paste0("plot for sample ", my_col_name, ".pdf"),
    ...) # other commands
}
my_plots <- lapply(my_sample_names, plot_one)
```

You could also use a `for` loop. 

> Also, can I make such separate graphs for a particular taxon such as Fungi only?

Yes, even easier. Use `filter_taxa` to subset the data before graphing. If you are only subsetting for the graph, you can use the `%>%` operator to avoid subsetting the data in the rest of your anaysis or making a copy of it. 

```r
obs %>%
  filter_taxa(taxon_names == "Funig", subtaxa = TRUE) %>%
  heat_tree(...)
```

**Comment by tarunaaggarwal** [comment]
Hey @zachary-foster... So I'm very close to getting the graph for only fungal groups but keep running into an `edge_label` error. 

```
obj = parse_qiime_biom("otu_table_mc2_w_tax_BlankOTUsRemoved_BlankSamplesRemoved_nem46.biom", 
                       class_regex = "^D?_?[0-9]*_?_?(.+)$")
print(obj)

obj$data$otu_table <- zero_low_counts(obj, "otu_table", min_count = 1)
no_reads <- obj$data$otu_table[, 1] == 0
sum(no_reads)


# Convert counts to proportions
obj$data$otu_table <- calc_obs_props(obj,
                                     dataset = "otu_table",
                                     cols = obj$data$sam_data$sample_ids)

# Calculate per-taxon proportions 
obj$data$otu_table <- calc_taxon_abund(obj, 
                                       dataset = "otu_table", 
                                       cols = obj$data$sam_data$sample_ids)


# construct heat tree
obj %>%
  filter_taxa(taxon_names == "Fungi", subtaxa = TRUE) %>%
  heat_tree(obj,
            node_size = n_obs,
            node_color = n_obs,
            node_label = taxon_names,
            tree_label = taxon_names)

```

Error:

```
Error in check_element_length(c("node_size", "edge_size", "node_label_size",  : 
  Length of argument'edge_size' must be a factor of the length of 'taxon_id'
```

I tried setting the `edge_size` to `taxon_names` but then I keep getting a similar error but for `edge_label` or `node_label_size` or `edge_label_size`, etc. 

Please let me know what I'm doing wrong here. Thank you so much!

**Comment by zachary-foster** [comment]
Hi @tarunaaggarwal, the error is because you gave the `obj` to the `heat_tree` command even though you already passed it in via the `%>%`, so it was trying to use the object for the next undefined parameter, which happened to be `edge_size`. Sorry for the unhelpful error message. Its a common enough mistake that I should make some code to look for it (#231).

 If you have not used `%>%` before: it takes what came before and uses it as the first input to the next function. For example, the two examples below do the exact same thing:

```r
obj %>%
  filter_taxa(taxon_names == "Fungi", subtaxa = TRUE) %>%
  heat_tree(node_size = n_obs,
            node_color = n_obs,
            node_label = taxon_names,
            tree_label = taxon_names)
```

```r
just_fungi <- filter_taxa(obj, taxon_names == "Fungi", subtaxa = TRUE)
heat_tree(just_fungi,
          node_size = n_obs,
          node_color = n_obs,
          node_label = taxon_names,
          tree_label = taxon_names)
```

**Comment by grabear** [comment]
@zachary-foster @tarunaaggarwal 

Hello guys.  From what I can tell the issue might not be with metacoder, but with the way you've parsed your data in phyloseq.  After taking a look at your .biom files that you linked above it looks like you've used the SILVA database to annotate your data.  Is that correct @tarunaaggarwal?

If you've used SILVA it designates ranks with the D_0__, D_1__, etc. prefixes.  (Unlike GreenGenes which has prefixes that look something like k_, p_, c_, etc.)

If so you can use the function that I've created and submitted as a PR to phylsoeq here:
https://github.com/joey711/phyloseq/pull/854

If you have the function loaded, then you'll end up doing something like this:
```r
# import biom data
silva_biom <- system.file("extdata", "SILVA_OTU_table.biom", package="phyloseq")
# Create phyloseq object using silva parseing function
silva_phyloseq <- import_biom(BIOMfilename = silva_biom, parseFunction = parse_taxonomy_silva_128)
```

**Comment by grabear** [comment]
I made a gist:
https://gist.github.com/grabear/018e86413b19b62a6bb8e72a9adba349

**Comment by tarunaaggarwal** [comment]
Hey @zachary-foster - Thank you for explaining the `%>%` usage. I have only used it a couple of times in the past and I keep forgetting that it operates like the `|`. It worked! Check out the lovely graph. Once I get the `for` loop to work, I will post my code in a separate issue page for anyone in the future who wishes to use it. 

Hi @grabear - I did indeed use the SILVA database. You caught that fast! 👍 Thank you for sending your R code. I'm sure it will be a huge time saver for our lab since we use Phyloseq a bunch. So does this mean you use SILVA database as well? How do you like the SILVA taxonomy for Qiime? We have been thinking about fixing the taxonomy manually but it is such a *HUGE* task. Just wondering what your opinion is. 

Thanks fellas! Appreciate your help! 


<img width="1003" alt="screen shot 2018-04-27 at 10 44 53 am" src="https://user-images.githubusercontent.com/9701326/39377478-24c2b4da-4a0a-11e8-80de-b31bed8bc842.png">

**Comment by grabear** [comment]
The microbiome project I'm working on is the first one I've done.  So it's also the first time I used metacoder.  But in my journey I read [this journal article](https://bmcgenomics.biomedcentral.com/articles/10.1186/s12864-017-3501-4).

I like that SILVA is more up to date (29/09/2016) vs GreenGenes (May 2013).

There's also this:

> SILVA, being the largest of the three 16S based taxonomies, shares the most taxonomic units with NCBI

That whole article was helpful to me.

**Comment by zachary-foster** [comment]
Thanks for you input @grabear!

@tarunaaggarwal:

Great! I am glad it worked. Anyone know where ascomycota and basidomycota are in SILVA's taxonomy? Seems like they should be there. 

> Once I get the for loop to work, I will post my code in a separate issue page for anyone in the future who wishes to use it.

Cool, thanks!

**Comment by tarunaaggarwal** [comment]
Hey @zachary-foster - I'm getting back to metacoder this week so I'm sorry I just noticed that you asked a question in your response. I just grep'd for D_5__Ascomycota and D_5__Basidiomycota in the taxonomy file that comes with SILVA (v132) and I found **3402** and **2443** OTUs containing **D_5__Ascomycota** and **D_5__Basidiomycota**, respectively, using the `consensus_taxonomy_all_levels.txt`. 

Were you using the file with 7 levels?

**Comment by grabear** [comment]
I believe Qiime uses the 7 level taxonomy file @tarunaaggarwal.  Not sure of other software that uses SILVA, so that's my only reference.

**Comment by tarunaaggarwal** [comment]
Hey @grabear - I thought you can specify any taxonomy file with any number of levels in Qiime. Either way, if I need more levels, I just replace the taxonomy strings with all 14 levels using a quick Python script. 

So how is Metacoder working out for you? I ask because I have been thinking of ways to make it work best for our lab. Ideally, I want to import the biom table (without taxonomy info), mapping file and a taxonomy file into metacoder without Phyloseq. Do you know how to do that?

**Comment by zachary-foster** [comment]
@tarunaaggarwal, I just ask because I did not see any fungal phyla in your plot and having them there would make it look better perhaps. Did they get filtered out before plotting?

**Comment by tarunaaggarwal** [comment]
Hey @zachary-foster! Oh I see. Those levels were not present in the 7 level taxonomy file we used to classify the OTUs I believe. Hence, they were missing. I have another question for you Zach. Is it possible to import the biom table (without taxonomy info), mapping file and a taxonomy file into metacoder without Phyloseq?

**Comment by zachary-foster** [comment]
So the biome file has an abundance matrix but no tax info? If you send me an example of each, I can probably tell you how to do it. Anything tabular for sure. I do have a biome parser as well, although I have never tried reading one without taxonomy info.

**Comment by tarunaaggarwal** [comment]
Thanks @zachary-foster ! [Here](https://www.dropbox.com/sh/narv17v5z7vxahf/AABEURI2FmWAEuZPvaesAETQa?dl=0) is the folder containing both types of biom tables - with and without tax. The one without taxonomy info is accompanied with a taxonomy file. I hope I'm not creating too much work for you. THANKS for your help!

**Comment by grabear** [comment]
Is there some reason that you need to use a biom file without taxonomies?

**Comment by tarunaaggarwal** [comment]
@grabear Sort of. We just reassigned taxonomy with SILVA 132 and I'd rather just find a way to work with the new taxonomy within R than to have to add to the biom table and refilter all over again. If its not feasible, its not the end of the world.

**Comment by zachary-foster** [comment]
@tarunaaggarwal, no problem! I found a way to do it: 


with taxonomy
-------------

``` r
library(metacoder)
```

``` r
with_tax <- parse_qiime_biom("biom-table-with-tax/otu_table_mc2_w_tax_BlankOTUsRemoved_BlankSamplesRemoved.biom")
```

``` r
print(with_tax)
```

    ## <Taxmap>
    ##   521 taxa: ab. D_0__Eukaryota ... ub. D_11__Norrlinia[truncated]
    ##   521 edges: NA->ab, ab->ac, ab->ad ... kd->tz, lq->ua, lr->ub
    ##   1 data sets:
    ##     otu_table:
    ##       # A tibble: 30,871 x 283
    ##         taxon_id otu_id   MEMB.nem.105 MEMB.nem.117 MEMB.nem.156
    ##         <chr>    <chr>           <dbl>        <dbl>        <dbl>
    ##       1 bh       FJ48040…           0.           0.           0.
    ##       2 ls       AF50812…           0.           0.           0.
    ##       3 lt       AY63045…           0.           0.           0.
    ##       # ... with 3.087e+04 more rows, and 278 more variables:
    ##       #   MEMB.nem.157 <dbl>, MEMB.nem.167 <dbl>,
    ##       #   MEMB.nem.168 <dbl>, MEMB.nem.176 <dbl>,
    ##       #   MEMB.nem.190 <dbl>, MEMB.nem.198 <dbl>,
    ##       #   MEMB.nem.200 <dbl>, MEMB.nem.22 <dbl>,
    ##       #   MEMB.nem.232 <dbl>, MEMB.nem.272 <dbl>, …
    ##   0 functions:

Without taxonomy
----------------

``` r
# Get OTU abundance
without_tax <- biomformat::read_biom("biom-table-without-tax/otu_table_mc2.biom")
```

    ## Warning in strsplit(msg, "\n"): input string 1 is invalid in this locale

``` r
otu_table <- dplyr::as_tibble(as.matrix(biomformat::biom_data(without_tax)))

# Get taxonomy file 
taxonomy <- readr::read_tsv("biom-table-without-tax/rep_set_tax_assignments.txt",
                            col_names = c("otu_id", "tax", "some_number"))
```

    ## Parsed with column specification:
    ## cols(
    ##   otu_id = col_character(),
    ##   tax = col_character(),
    ##   some_number = col_double()
    ## )

``` r
# Combine both in a taxmap object 
obj <- parse_tax_data(taxonomy,
                      class_cols = "tax", class_sep = ";",
                      datasets = list(otu_table = otu_table),
                      mappings = c("{{index}}" = "{{index}}"))
print(obj)
```

    ## <Taxmap>
    ##   2375 taxa: aab. D_0__Eukaryota ... dnj. D_14__
    ##   2375 edges: NA->aab, aab->aac ... czz->dni, daa->dnj
    ##   2 data sets:
    ##     tax_data:
    ##       # A tibble: 31,848 x 4
    ##         taxon_id otu_id    tax                       some_number
    ##         <chr>    <chr>     <chr>                           <dbl>
    ##       1 dab      New.Refe… D_0__Eukaryota;D_1__Opis…       0.700
    ##       2 amc      New.Refe… D_0__Eukaryota;D_1__Opis…       0.820
    ##       3 amd      New.Clea… D_0__Eukaryota;D_1__Opis…       1.00 
    ##       # ... with 3.184e+04 more rows
    ##     otu_table:
    ##       # A tibble: 31,848 x 305
    ##         taxon_id MEMB.nem.105 MEMB.nem.117 MEMB.nem.156
    ##         <chr>           <dbl>        <dbl>        <dbl>
    ##       1 dab                4.           1.          36.
    ##       2 amc                0.           0.           0.
    ##       3 amd                0.           0.           0.
    ##       # ... with 3.184e+04 more rows, and 301 more variables:
    ##       #   MEMB.nem.157 <dbl>, MEMB.nem.167 <dbl>,
    ##       #   MEMB.nem.168 <dbl>, MEMB.nem.176 <dbl>,
    ##       #   MEMB.nem.190 <dbl>, MEMB.nem.198 <dbl>,
    ##       #   MEMB.nem.200 <dbl>, MEMB.nem.22 <dbl>,
    ##       #   MEMB.nem.232 <dbl>, MEMB.nem.272 <dbl>, …
    ##   0 functions:

If you want to remove the "D_0__" and things like "(animals)" from the names, you can use the `class_key` and `class_regex` options to do that if you know some regex

**Comment by tarunaaggarwal** [comment]
@zachary-foster Nice! I will try this right away. And how to import the mapping file please?

**Comment by zachary-foster** [comment]
Oh yea, the mapping file. It has no taxonomic info associated with it, so it can stay a separate table. The `taxmap` object is only concerned with data that has taxonomic info associated with it, unlike `phyloseq` objects. You could put the mapping file in there, but it would not help anything. I like the `readr` package for tabular data:

``` r
mapping <- readr::read_tsv("18S-euk-QIIME-mapping-MEmicrobiome-FINAL-26Jul17.txt")
```

If you want, you could add it to the `taxmap` object like so: 

```r
obj$data$mapping <- readr::read_tsv("18S-euk-QIIME-mapping-MEmicrobiome-FINAL-26Jul17.txt")
```

`obj$data` is a list, so you can put anything you want there, but it will not make things easier unless what you put there is named by taxon IDs.

**Comment by tarunaaggarwal** [comment]
Morning @zachary-foster - may I please have your email address?

**Comment by zachary-foster** [comment]
Sure, its `zacharyfoster1989` a `gmail.com`

**Comment by tarunaaggarwal** [comment]
Thanks! I emailed ya.

**Comment by zachary-foster** [comment]
Closing due to inactivity.  If there are still unresolved issues, feel free to reopen this issue or open a new issue.

**Comment by sunilmundra** [comment]
Hi @zachary-foster
I am trying to make a differential heat tree for two categories (birch vs spruce)
Everything works fine till I plot overall tree, tree for individual categories, but when I am planning to make differential tree for comparing two categories based on log2 mean ration I am getting error (see code and error below).
###################
# Differential tree
###################
metacoder.data$taxon_names
#first calculate the differences in abundances based on the two groupings
families$data$diff_table <- compare_groups(metacoder.data, data = "tax_by_host", #_by_host OR specify which information to use - here it is the abundance of the reads - you could also do the proportion in the samples using tax_data instead
                                           cols = c("Birch","Spruce"), #specify the names of the columns containing your OTU counts
                                           groups = c("Birch","Spruce"))
families$data$diff_table$log2_median_ratio
heat_tree(families,
          node_label = taxon_names, #specifies what names to give the circles (nodes)
          node_size = n_obs, #size nodes by total number of reads
          node_color = log2_median_ratio, #specifies the differences between groups
          #edge_color = log2_median_ratio,
          #edge_size = total,#thickness of lines determined by total # of reads
          node_color_interval = c(-2, 2),
          node_color_range = c("cadetblue", "grey75", "darkorange1"),
          node_label_size_range = c(0.007,0.06),#adjust the min-max numbers to change the relative size of the text
          node_size_axis_label = "Total Taxon Abundance",
          node_color_axis_label = "Differentially Occurring Samples")

> Error in check_element_length(c("node_size", "edge_size", "node_label_size",  : 
>   Length of argument'node_color' must be a factor of the length of 'taxon_id'

There seems some problems in defining node_color, and I am not able to figure out how should I get it correct?

Looking forward to hear from you

Regards
Sunil

**Comment by zachary-foster** [comment]
Hi Sunil, 

What does the print out for `families` look like before plotting? Thanks

**Comment by ChathuZ** [comment]
When I convert the phyloseq to tax map its very different. Any suggestions to proceed from here?
[phyloseq_to_metacoder.docx](https://github.com/user-attachments/files/18172636/phyloseq_to_metacoder.docx)
